content/posts/backlogged-pinball.md

```diff
@@ -0,0 +1,99 @@
++++
+title =  "Backlogged Pinball"
+tags = ["pinball", "gcp", "next"]
+date = "2024-10-07"
++++
+
+# Backlogged Pinball
+
+Earlier this year, I built a pinball-based demo for [Cloud Next 2024](https://cloud.withgoogle.com/next).  I'm preparing some posts about specific aspects of it, but I wanted to post a brief overview first.  Backlogged is a physical pinball machine that has custom software to connect it to Cloud
+services.
+
+
+{{< figure src="/images/backlogged-pinball/pinball_suit.jpeg" width="600px" title="A conference attendee playing the Backlogged pinball demo" >}}
+
+## Demo overview
+
+
+
+Backlogged pinball was designed to demonstrate specific cloud products and
+benefits in a fun, interactive way.
+
+{{< figure src="/images/backlogged-pinball/backlogged-arch.png" width="600px" title="The pinball machine sends pubsub events to the cloud which are processed by Cloud Run and displayed on an interactive web app. Users can deploy their own Cloud Run service to process events and send pubsub messages back to the machine." >}}
+
+In its current iteration, it highlights:
+
+1.  Cloud PubSub: Two-way PubSub communication with a novel source of events
+    (The pinball machine)
+1.  Cloud Run / Cloud functions: Both serverless platforms process pinball
+    events in real-time using different subscription mechanics (push vs. pull).
+1.  Cloud Run Deployments: Users can update a Cloud Run service in 20 seconds
+    which has an immediate effect on the current pinball game
+1.  Cloud Firestore: Firestore pushes all events to a live web-app. The round
+    trip from:
+
+    Pinball machine **->** PubSub **->** Cloud Functions **->** Firestore **->**
+    Web App
+
+    takes about 200 ms, and could probably be further optimized.
+
+## The machine
+
+{{< figure src="/images/backlogged-pinball/pinball_wide.jpeg" width="600px" title="The demo featured live-updating high scores and current game status" >}}
+
+The actual physical pinball machine is a
+[Multimorphic P3](https://www.multimorphic.com/p3-pinball-platform/). In short,
+this is a modular, programmable pinball machine with 2 large HD displays built
+in.
+
+Multimorphic also distributes
+[an SDK](https://www.multimorphic.com/p3-pinball-platform/3rd-party-developers/)
+to facilitate creating new pinball games.
+
+Thus, Backlogged is an application that uses the Unity game engine and the
+Multimorphic SDK to handle interfacing with the hardware (including displays).
+
+## What's next?
+{{< figure src="/images/backlogged-pinball/pinball_interview.jpeg" width="600px" title="Drew showing off the backlogged demo" >}}
+
+The pinball machine currently resides in a Google office in Salt Lake City.  Whenever I am able to, I plan on adding new features/functionality to the demo (with a major focus on cloud code).
+
+I am currently working on a game advisor that uses [llama3.1-70b](https://ollama.com/library/llama3.1:70b) to grade your game and recommend more optimal strategies. 
```

Comment:
is this based on the game or on pinball itself? like does it give recommendations on how to be better at pinball? regardless this is fascinating and makes me want to see more :)

Reply:
It gives some generic pinball tips (pretty decent ones too) and it's getting better at giving game-specific advice.  Stay tuned for further updates closer to Kubecon!